### PR TITLE
drivers/flash/nrf_qspi_nor: Set custom QE bit

### PR DESCRIPTION
**Pull request by natelust**
Modify the nordic flash driver to allow specifying which bit
corresponds to the quad enable bit. This is needed because not
all NOR chips conform to having QE on the 6th bit, as was hard
coded. Now it is possible to set any bit in the first register, as
well as bit a second register if the flash chip needs.

This is motivated by the adafrut feather nrf52840 board being unable
to write to the on-board "external" flash using existing DT bindings.

### CODE REVIEW

**Comment by natelust** [comment]
This corresponds to issue: https://github.com/zephyrproject-rtos/zephyr/issues/42333

**Comment by natelust** [comment]
I should note I built this patch against the current zephyr 2.7 head, and future ported it to 3.0 (which was straightforward to apply), but I need to build the new image with 3.0 and test it on my board tonight. It works on 2.7, and I don't see why it wouldn't but I need to check, so extra close look at this is very welcome (especially as my c is kind of rusty).

**Comment by de-nordic** [comment]
The position of the QE bit is standardized by JEDEC 216D (page 40) and is limited to only several bits as the 6.4.18 JEDEC BFP 15th DWORD Table shows. AF board used GD25 which uses 2 status registers and the current code is for MX25 which uses single one.
I think that rather than directly pointing to the bit that enables the feature, the `quad-enable-bit` (or bits) should directly take values that are allowed to the BFP table 15th dword, for example `quad-enable-bits = 000b`, QE not supported, `quad-enable-bits = 001b`, QE in 1st bit of SR#2, etc.

There is also DTS support for the SDFP and run-time SDFP that is probably not addressed.

**Comment by nvlsianpu** [comment]
@de-nordic  @natelust ^^ here ticked I had created for tracking this missing feature. Expected solution is described under it.

**Comment by natelust** [comment]
> The position of the QE bit is standardized by JEDEC 216D (page 40) and is limited to only several bits as the 6.4.18 JEDEC BFP 15th DWORD Table shows. AF board used GD25 which uses 2 status registers and the current code is for MX25 which uses single one. I think that rather than directly pointing to the bit that enables the feature, the `quad-enable-bit` (or bits) should directly take values that are allowed to the BFP table 15th dword, for example `quad-enable-bits = 000b`, QE not supported, `quad-enable-bits = 001b`, QE in 1st bit of SR#2, etc.
> 
> There is also DTS support for the SDFP and run-time SDFP that is probably not addressed.

@de-nordic I am happy for whatever you feel is best, I am just some random software person who had no knowledge about any of this until last week when I started going over code and docs to try to get this working. I am not part of a company, so I could not figure out how to get a hold of the JEDEC spec (Registration seems to want some affiliation) and had to guess at some areas, and perhaps the proposed solution is more general than it needs to be.

Forgive me if you understand all this already, I am just trying to clarify for my understanding. The solution I proposed assumes 16 bits where the first 8 are assumed to be in the first register, and if a higher bit number is specified then it assumes the QE bit must be in second register, encoding both pieces of info together. Are you saying the spec allows for more than one bit needing to be set? Is your proposal that `quad-enable-bits` to be given as the direct bitmask to apply, instead of what the code currently does of creating the mask from the hard coded bit 6? I guess not much would else would need to change, as we could still detect if the second register needs to be read/written and emit the appropriate code, assuming it was valid for this bit pattern to still be 16 bits.

I would be happy to make that change if you wanted to see what it looked like (or think it is the right solution). Would you be able to supply me with what bit patterns are allowed per the spec, so I can appropriately modify the enum in the binding?

@nvlsianpu I apologize, I am not sure I follow what you are saying

File: dts/bindings/mtd/nordic,qspi-nor.yaml
Comment (anangl) [inline dts/bindings/mtd/nordic,qspi-nor.yaml:32]: A property for specifying this bit already exists, see [quad-enable-requirements](https://github.com/zephyrproject-rtos/zephyr/blob/83939c267c1a4ffe2644a913da79b5409e114f81/dts/bindings/mtd/jedec%2Cjesd216.yaml#L41-L58). The nrf_qspi_nor driver currently supports only the `S1B6` value that is [set as default](https://github.com/zephyrproject-rtos/zephyr/blob/83939c267c1a4ffe2644a913da79b5409e114f81/dts/bindings/mtd/nordic%2Cqspi-nor.yaml#L26-L28) for `"nordic,qspi-nor"` nodes.
#42049 adds support for the `S2B1v4` value. It seems that it should work for GD25Q16C, too.

**Comment by natelust** [comment]
This seems to already be fixed in PR https://github.com/zephyrproject-rtos/zephyr/pull/42049. Thanks for your time looking at this.

**Comment by natelust** [comment]
It turns out the other ticket DOES not fix this issue, my test board happened to have the QE bit set when I opened it. However, I think some of this work can be modified in context of the other patch. I am reopening this to keep track of the fact that this is still broken.

**Comment by nvlsianpu** [comment]
> @nvlsianpu I apologize, I am not sure I follow what you are saying

I missed DW16 with DW15. Forget about my comment.

File: dts/bindings/mtd/nordic,qspi-nor.yaml
Comment (mbolivar-nordic) [inline dts/bindings/mtd/nordic,qspi-nor.yaml]: I will leave it up to the QSPI maintainers at Nordic whether or not to accept this change, but from the devicetree side, the use of `default` without a justification here is not allowed.

Please read https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings and update this accordingly; thanks.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: dts/bindings/mtd/nordic,qspi-nor.yaml
Comment (carlescufi) [reply dts/bindings/mtd/nordic,qspi-nor.yaml]: @natelust could you please address this comment?

Comment (natelust) [reply dts/bindings/mtd/nordic,qspi-nor.yaml]: Sorry there is two similar pull requests going on. This work has been mostly subsumed by the other pr, so I am closing this one.

**Pull request closed without merging**
